fix(Earthfile): Build also the go binaries when running +build

tampakrap opened the pull request:
<!--
Thank you for helping to improve Crossplane! Please read the contribution docs
(linked below) if this is your first Crossplane pull request.
-->

### Description of your changes

The `+multiplatform-build` step includes `+go-multiplatform-build`, so similarly the `build` step should include `+go-build`.

When calling the `+go-build` target, the OS/arch that is used is the native, but in some cases, eg on Docker Desktop on Mac OS X, emulation is used, thus the $NATIVEPLATFORM and $TARGETPLATFORM variables incorrectly report linux/amd64. The USERPLATFORM var though always returns the correct OS/arch of the host that is triggering the build.

When calling the `+build` target locally, the expectation is to produce a binary for our own system, so that we can use/test it afterwards.

Thus, this commit changes the `+build` target to call `+go-build` but for platform=$USERPLATFORM.

Addresses the feedback from https://github.com/crossplane/crossplane/issues/5750#issuecomment-2151204269

<!--
Briefly describe what this pull request does, and how it is covered by tests.
Be proactive - direct your reviewers' attention to anything that needs special
consideration.

We love pull requests that fix an open issue. If yours does, use the below line
to indicate which issue it fixes, for example "Fixes #500".
-->

I have: <!--You MUST either [x] check or [ ] ~strike through~ every item.-->

- [X] Read and followed Crossplane's [contribution process].
- [X] Run `earthly +reviewable` to ensure this PR is ready for review.
- [ ] ~Added or updated unit tests.~
- [ ] ~Added or updated e2e tests.~
- [ ] ~Linked a PR or a [docs tracking issue] to [document this change].~
- [ ] ~Added `backport release-x.y` labels to auto-backport this PR.~

Need help with this checklist? See the [cheat sheet].

[contribution process]: https://github.com/crossplane/crossplane/tree/master/contributing
[docs tracking issue]: https://github.com/crossplane/docs/issues/new
[document this change]: https://docs.crossplane.io/contribute/contribute
[cheat sheet]: https://github.com/crossplane/crossplane/tree/master/contributing#checklist-cheat-sheet

jbw976 reviewed (approved):
This looks good to me @tampakrap, I think it's totally reasonable to build the go binaries during the `earthly +build` target 👍 

I tested this out and I see the binaries now in the `_output` tree:
```
❯ tree _output
_output
├── bin
│   └── linux_arm64
│       ├── crank
│       └── crossplane
└── charts
    └── crossplane-0.0.0-1717663449-13c28b15.tgz

3 directories, 3 files
```

One curiosity question - do you already know the proper way to build a specific platform's binaries? e.g. I'm running on Mac M1 with Apple silicon, so I'd like to have the `darwin/arm64` binaries to run locally. I tried the command below but still got `linux/arm64` - have you already figured this out? 😇 
```
earthly +build --platform=darwin/arm64
```

jbw976 commented:
It'd be cool to merge this one first so it helps with testing your other PRs:

* https://github.com/crossplane/crossplane/pull/5779
* https://github.com/crossplane/crossplane/pull/5781

tampakrap commented:
> One curiosity question - do you already know the proper way to build a specific platform's binaries? e.g. I'm running on Mac M1 with Apple silicon, so I'd like to have the `darwin/arm64` binaries to run locally. I tried the command below but still got `linux/arm64` - have you already figured this out? 😇
> 
> ```
> earthly +build --platform=darwin/arm64
> ```

It kinda works if you do `earthly --platform=darwin/arm64 +build`, the problem is that the +image step will fail because it supports only linux arch. The best way to achieve it is to use `earthly --platform=darwin/arm64 +go-build`. Alternatively, we could skip the `+image` and `+helm-chart` steps, or override the target platform to linux/amd64. Either way, it is out of scope of this PR, as this is the behavior in master currently

negz commented:
I wonder - what's the point of the `+build` target? How do folks use it?

It's not used in CI, because we use the multiplatform build there. I assume it's only used for local development. Do folks really want a target that builds an image, binaries, and Helm chart? Are they going to use all three at once for something?

tampakrap commented:
I agree, I believe `+build` should just call `+go-build`. But also Jared has a point here, the expected behavior would have been to produce a crank binary for my system's OS/arch, so I can also test it afterwards

tampakrap commented:
I got it, the issue is that the docker desktop is using emulation on Mac, so it is setting the NATIVEPLATFORM USERPLATFORM and TARGETPLATFORM env vars incorrectly to linux/amd64. I introduced a new target called `+go-build-locally` in this PR, please have a look and give me some feedback.

If you like it then I suggest to put it in the `+build` target and to remove the image and helm ones from there.

jbw976 commented:
> what's the point of the +build target? How do folks use it?

I've always looked at it from the perspective of someone who doesn't know all the capabilities of the build system: "I don't know what all artifacts you're going to build, just build all of them for my local platform please".

I've always appreciated the main `build` target to build all the main artifacts for me - even if I don't know what they are. If I want to build something specific then I can do so, but if I'm not specific then just build them all. That enables folks to:

* know the build is working OK for all the main artifact paths, without having to figure out what those all are and call each one of them
* have all the artifacts available to use locally, if they choose to

jbw976 commented on Earthfile:
```diff
@@ -159,6 +160,16 @@ go-build:
+go-build-locally:
+  LOCALLY
+  ARG TARGETPLATFORM
+  BUILD --platform=$TARGETPLATFORM +go-build
```
i like this being the default for `earthly +build`, so that plain vanilla command will result with binaries for the host system that invoked the build 🤩

tampakrap replied:
There is a limitation though, when using `LOCALLY` we can not override the platform, so in a command similar to this `earthly --platform=windows/amd64 +go-build-locally` the --platform argument will be ignored.

Thus, the workflow for building locally the go binaries becomes the following:

1. For binaries with the same OS/arch as the host: `earthly +go-build-locally`
1. For binaries with the OS/arch as the emulation used in the host: `earthly +go-build`
1. For specifying the OS/arch: `earthly --platform=$OS/$ARCH +go-build`
1. For binaries on all available OS/archs: `earthly +go-multiplatform-build`

So if we add +go-build-locally in +build, it will become the same as 1. Are we fine with that?

negz replied:
I don't think `LOCALLY` is doing anything here, is it?

`LOCALLY` means "run this build in my local environment, not in a container", but `BUILD` is going to run the `go-build` target in a container.

jbw976 replied:
ah thanks for clearing that up about the `LOCALLY` directive @negz - I agree that it would be best to continue running a containerized build, but do cross compilation when needed.

tampakrap replied:
correct, the `+go-build-locally` target was just setting the ARG and then passing the build to `go-build` which performs the actual build in a container. But either way, since the USERPLATFORM solution from above seems to work, I'm happily removing this target

tampakrap commented:
I added `+go-build-locally` in the `+build` target, and also I updated the commit message and the description of this PR to mention the limitation and the local build workflow options as described in [this comment](https://github.com/crossplane/crossplane/pull/5782#discussion_r1630811537).

negz commented on Earthfile:
```diff
@@ -25,6 +25,7 @@ lint:
 
 # build builds Crossplane for your native OS and architecture.
 build:
+  BUILD +go-build-locally
```
```suggestion
  ARG USERPLATFORM
  BUILD +go-build --platform=${USERPLATFORM}
```

Would this achieve the goal of causing Earthly to build binaries for the user's platform (e.g. Darwin)?

It works for me, but I run Earthly on Linux so for me USERPLATFORM == NATIVEPLATFORM.

jbw976 replied:
I believe I have tested this suggestion out on my Mac M1 but it didn't seem to work for me. I synced to latest from https://github.com/tampakrap/crossplane/tree/earthfile_build_step_also_build_go, applied this suggestion on top, and ran `earthly +build`.

I do see build output that looks correct: `+go-build | platform=darwin/arm64`, but the `_output` dir still only has `_output/bin/linux_arm64/`.

A few more details about my testing are available in the following gist: 
https://gist.github.com/jbw976/d7a3d230b5cfaefb4259f50dbe056c4a

jbw976 replied:
aha, if I just pass the `--platform` arg before the `+go-build` target, then it does work as intended! e.g.:

```
BUILD --platform=${USERPLATFORM} +go-build
```

Gives me:
```
❯ tree _output
_output
├── bin
│   └── darwin_arm64
│       ├── crank
│       └── crossplane
└── charts
    └── crossplane-0.0.0-1717748817-e9886233.tgz

3 directories, 3 files
```

tampakrap replied:
@negz you are absolutely right! I tried the following and I could swear that it showed linux in all cases, but now I tried it again and indeed USERPLATFORM shows the correct one! Many thanks for spotting this and ultimately providing the correct solution!

```Dockerfile
# theo is testing stuff
theo:
  ARG NATIVEPLATFORM
  ARG USERPLATFORM
  ARG TARGETPLATFORM
  FROM nginx:alpine
  RUN echo "NATIVEPLATFORM is $NATIVEPLATFORM"
  RUN echo "USERPLATFORM is $USERPLATFORM"
  RUN echo "TARGETPLATFORM is $TARGETPLATFORM"
```

```
               +theo | --> RUN echo "NATIVEPLATFORM is $NATIVEPLATFORM"
               +theo | NATIVEPLATFORM is linux/amd64
               +theo | --> RUN echo "USERPLATFORM is $USERPLATFORM"
               +theo | USERPLATFORM is darwin/amd64
               +theo | --> RUN echo "TARGETPLATFORM is $TARGETPLATFORM"
               +theo | TARGETPLATFORM is linux/amd64
```

negz reviewed (changes requested):
I don't think we should be building anything `LOCALLY`, it defeats the purpose of a containerized build. It should be possible to cross-compile binaries for Darwin inside a container.

negz commented:
> I got it, the issue is that the docker desktop is using emulation on Mac, so it is setting the NATIVEPLATFORM USERPLATFORM and TARGETPLATFORM env vars incorrectly to linux/amd64.

@tampakrap What version of Earthly are you running? My understanding is `USERPLATFORM` should be `darwin/arm64` when running the `earthly` binary on Apple Silicon in MacOS. It sounds like before v0.6.29 a bug caused it to be incorrect - https://github.com/earthly/earthly/issues/2446.

jbw976 reviewed (commented):
From my testing/troubleshooting with @negz suggestion, we can probably get rid of the `+go-build-locally` target and simply pass `--platform=${USERPLATFORM}` to `+go-build` from within the `+build` target, so `earthly +build` will build binaries for the user's local platform, along with the container images and helm chart.

That seems like a sufficient experience for my usage at least 😉

tampakrap commented:
many thanks @negz @jbw976 I applied the change and removed the commits that introduced the `+go-build-locally` please give it a review one more time

jbw976 commented on Earthfile:
```diff
@@ -25,6 +25,8 @@ lint:
 
 # build builds Crossplane for your native OS and architecture.
 build:
+  ARG USERPLATFORM
```
this looks correct to me with my testing of `early +build` now (i get `darwin_arm64` binaries):
```
❯ tree _output
_output
├── bin
│   └── darwin_arm64
│       ├── crank
│       └── crossplane
└── charts
    └── crossplane-0.0.0-1717875335-2e37d912.tgz

3 directories, 3 files
```

jbw976 reviewed (approved):
I like where this landed with the functionality and the simplicity 🤓 

Consider rebasing these commits to just a single commit, otherwise LGTM!

tampakrap commented:
> Consider rebasing these commits to just a single commit, otherwise LGTM!

Done
